File: 946.验证栈序列.py
#
# @lc app=leetcode.cn id=946 lang=python3
#
# [946] 验证栈序列
#

# @lc code=start
class Solution:
    def validateStackSequences(self, pushed: List[int], popped: List[int]) -> bool:
        stack = []
        pop_index = 0
        for num in pushed:
            # get in stack
            stack.append(num)
            while stack and stack[-1] == popped[pop_index]:
                # a element get out of the stack and move pop_index
                stack.pop(-1)
                pop_index += 1
        # if ok, then the stack is empty
        return not stack
# @lc code=end

File: 946.验证栈序列.go
/*
 * @lc app=leetcode.cn id=946 lang=golang
 *
 * [946] 验证栈序列
 */

// @lc code=start
func validateStackSequences(pushed []int, popped []int) bool {
	stack := make([]int, 0, len(pushed))
	popIndex := 0
	for _, num := range pushed {
		// get in stack
		stack = append(stack, num)
		// check if current situation could work,
		// yes one element out stack,
		// no continue(the pop one may be the next one in pushed)
		for len(stack) != 0 && stack[len(stack)-1] == popped[popIndex] {
			stack = stack[:len(stack)-1]
			popIndex++
		}
	}
	return len(stack) == 0
}

// @lc code=end

